README
NAME
    Net::Squid::Auth::Plugin::SimpleLDAP - A simple LDAP-based credentials
    validation plugin for Net::Squid::Auth::Engine

VERSION
    Version 0.1.80

SYNOPSIS
    If you're a system administrator trying to use Net::Squid::Auth::Engine
    to validate your user's credentials using a LDAP server as a credentials
    repository, do as described here:

    On "$Config{InstallScript}/squid-auth-engine"'s configuration file:

      plugin = SimpleLDAP
      <SimpleLDAP>
        # LDAP server
        server = myldap.server.somewhere       # mandatory

        # connection options
        <NetLDAP>                              # optional section with
          port = N                             #   Net::LDAP's
          scheme = 'ldap' | 'ldaps' | 'ldapi'  #     constructor
          ...                                  #     options
        </NetLDAP>

        # bind options
        binddn = cn=joedoe                     # mandatory
        bindpw = secretpassword                # mandatory

        # search options
        basedn = ou=mydept,o=mycompany.com     # mandatory
        objclass = inetOrgPerson               # opt, default "person"
        userattr = uid                         # opt, default "cn"
        passattr = password                    # opt, default "userPassword"
      </SimpleLDAP>

    Unless configured otherwise, this module will assume the users in your
    LDAP directory belong to the object class "person", as defined in
    section 3.12 of RFC 4519, and the user and password information will be
    looked for in the "cn" and "userPassword" attributes, respectively.
    Although you can choose to use any other pair of attributes, the
    "userattr" can be set to "DN", while the "passattr" can not.

    On your Squid HTTP Cache configuration:

        auth_param basic /usr/bin/squid-auth-engine /etc/squid-auth-engine.conf

    And you're ready to use this module.

    If you're a developer, you might be interested in reading through the
    source code of this module, in order to learn about it's internals and
    how it works. It may give you ideas about how to implement other plugin
    modules for Net::Squid::Auth::Engine.

FUNCTIONS
  new( $config_hash )
    Constructor. Expects a hash reference with all the configuration under
    the section *<SimpleLDAP>* in the
    "$Config{InstallScript}/squid-auth-engine" as parameter. Returns a
    plugin instance.

  initialize()
    Initialization method called upon instantiation. This provides an
    opportunity for the plugin initialize itself, stablish database
    connections and ensure it have all the necessary resources to verify the
    credentials presented. It receives no parameters and expect no return
    values.

  _search()
    Searches the LDAP server. It expects one parameter with a search string
    for the username. The search string must conform with the format used in
    LDAP queries, as defined in section 3 of RFC 4515.

  is_valid( $username, $password )
    This is the credential validation interface. It expects a username and
    password as parameters and returns a boolean indicating if the
    credentials are valid (i.e., are listed in the configuration file) or
    not.

  config( $key )
    Accessor for a configuration setting given by key.

AUTHOR
    Alexei Znamensky, "<russoz at cpan.org>"

BUGS
    Please report any bugs or feature requests to
    "bug-net-squid-auth-plugin-simpleldap at rt.cpan.org", or through the
    web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Net-Squid-Auth-Plugin-Si
    mpleLDAP>. I will be notified, and then you'll automatically be notified
    of progress on your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Net::Squid::Auth::Plugin::SimpleLDAP

    Or take a look at the github site to be up to date:

        <http://github.com/russoz/Net-Squid-Auth-Plugin-SimpleLDAP>

    You can also look for information at:

    *   RFC 4515 - Lightweight Directory Access Protocol (LDAP): String
        Representation of Search Filters

        <http://www.faqs.org/rfcs/rfc4515.html>

    *   RFC 4519 - Lightweight Directory Access Protocol (LDAP): Schema for
        User Applications

        <http://www.faqs.org/rfcs/rfc4519.html>

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=Net-Squid-Auth-Plugin-Simp
        leLDAP>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/Net-Squid-Auth-Plugin-SimpleLDAP>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/Net-Squid-Auth-Plugin-SimpleLDAP>

    *   Search CPAN

        <http://search.cpan.org/dist/Net-Squid-Auth-Plugin-SimpleLDAP>

SEE ALSO
    Net::Squid::Auth::Engine, Net::LDAP

ACKNOWLEDGEMENTS
    Luis "Fields" Motta Campos "<lmc at cpan.org>", who could now say:

    "The circle is now complete. When I left you, I was but the learner; now
    *I* am the master."

    To what I'd reply:

    "Only a master of Perl, Fields"

COPYRIGHT & LICENSE
    Copyright 2008,2010 Alexei Znamensky, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Net-Squid-Auth-Plugin-SimpleLDAP

0.1.82  [2010.11.08 07:45:43] Net::Squid::Auth::Plugin::SimpleLDAP DIST VERSION 0.1.82
======= ==============================================================================
1178eaa refrained from using op //. using plain || instead, and allowing older perls
5549025 added build files by mistake - removing now

0.1.81  [2010.11.07 21:01:38] Net::Squid::Auth::Plugin::SimpleLDAP DIST VERSION 0.1.81
======= ==============================================================================
bb91025 added build files by mistake - removing now
64ac94a using Scalar::Util to test for hash
0361349 mkdist output formatting
1d076c7 improved mkdist further still
2f4305c Improved mkdist script

0.1.80  13.10.2010 Net::Squid::Auth::Plugin::SimpleLDAP DIST VERSION 0.1.80
======= ======================================================================
f21d056 improved mkdist
d785e42 Run test through perltidy
31d00d0 Improved documentation about using dn as user

0.1.16  15.09.2010 DIST VERSION 0.1.16
======= ======================================================================
02fd22b Messed up versions. Moving up to 0.0.16.
aed46ce mkdist: print location of finished package for convenience

0.1.15  15.09.2010 DIST VERSION 0.1.15
======= ======================================================================
e5d7881 Added dry-run notice to the end of the script
ea5498d Eye candy in the Changelog generation
af59f77 Fixed the case where the user attribute is the DN
a61db8b More reasonable parameter checking in new()
1f70681 Reduced one-level in the directory tree

0.1.14  10.03.2010 DIST VERSION 0.1.14
e9d43af Just run t/perlcritic.t if Russoz is in the house!

0.1.13  06.03.2010 DIST VERSION 0.1.13
bf1c0b8 mkdist: Changed format of output into Changes file.
e85a02a Created accessor method config(). Changed code to use it.
7067349 Added perltidy to mkdist. Trimmed unnecessary check in git log output.
791e6a1 Removed META.yml again. Fixed mkdist, so it won't come back.
91eddfd No longer using "my $_" due to older versions of perl
425105e Not adding META.yml to git
ff83008 The Perl Critic test was poorly written.
6d8be1c Fixed the github URL.
3a0db0d Yet more candy on dev/mkdist. Should I make a module out of it?
988be4d No need to keep version of this build-generated file
4d0fb8b Candy candy I won't let you go..
06ebfd9 This is perl, son, not bash.

0.1.12  04.03.2010
        be56c30 Not allowing the same version to be commited twice. Printing first 20 lines of Changes.

0.1.11  04.03.2010

0.1.10  04.03.2010
        8affd80 Dry-run mode by default. Use -f to force. Adding module to git too.
        7ad5af7 Small fix in the -n code
        a294da4 Better messages. Added a '-n' switch for dry-run.
        1731432 Removed Module::Install stuff from git

0.1.9   04.03.2010
        45f857a Added messages and error handling to mkdist
        229719a Fixed mkdist to properly hold the new files
        21fa5a6 Removed old Module::Install stuff. Long live Module::Build!
        62fb076 mkdist did not added the updated Changes file

0.1.8   04.03.2010
        0b2a3d7 mkdist script is now ready for use
        0418c7d Fixed github URL in Changes
        ffd5423 Param validations in new(). initialize() for connecting

0.01.06 04.03.2010
        Changed the name of the Net::LDAP options section.
        More TO DO. Upgraded VERSION.
        Ran through perltidy

0.01.05 04.03.2010
        Fixed the module abstract.
        Added t/perlcritic.t anf fixed MANIFEST

0.01.04 03.03.2010
        Moved to github. See more at:
        http://github.com/russoz/Net-Squid-Auth-Plugin-SimpleLDAP
        Updated README.
        Added more perldoc lines.
        Moved to Module::Build.
        Added some tests.
        Now allows config file to hold options for Net::LDAP.

0.01.03 14.08.2009
        Public release

0.01    Date/time
        First version, released on an unsuspecting world.



lib/Net/Squid/Auth/Plugin/SimpleLDAP.pm
package Net::Squid::Auth::Plugin::SimpleLDAP;

use warnings;
use strict;
use Carp;
use Net::LDAP;
use Scalar::Util qw/reftype/;

=head1 NAME

Net::Squid::Auth::Plugin::SimpleLDAP - A simple LDAP-based credentials validation plugin for Net::Squid::Auth::Engine

=cut

use version; our $VERSION = qv('0.1.82');

=head1 SYNOPSIS

If you're a system administrator trying to use Net::Squid::Auth::Engine to
validate your user's credentials using a LDAP server as a credentials
repository, do as described here:

On C<$Config{InstallScript}/squid-auth-engine>'s configuration file:

  plugin = SimpleLDAP
  <SimpleLDAP>
    # LDAP server
    server = myldap.server.somewhere       # mandatory

    # connection options
    <NetLDAP>                              # optional section with
      port = N                             #   Net::LDAP's
      scheme = 'ldap' | 'ldaps' | 'ldapi'  #     constructor
      ...                                  #     options
    </NetLDAP>

    # bind options
    binddn = cn=joedoe                     # mandatory
    bindpw = secretpassword                # mandatory

    # search options
    basedn = ou=mydept,o=mycompany.com     # mandatory
    objclass = inetOrgPerson               # opt, default "person"
    userattr = uid                         # opt, default "cn"
    passattr = password                    # opt, default "userPassword"
  </SimpleLDAP>

Unless configured otherwise, this module will assume the users in your LDAP 
directory belong to the object class C<person>, as defined in section 3.12 of 
RFC 4519, and the B<user> and B<password> information will be looked for in the 
C<cn> and C<userPassword> attributes, respectively. Although you can choose
to use any other pair of attributes, the C<userattr> can be set to C<DN>,
while the C<passattr> can not.

On your Squid HTTP Cache configuration:

    auth_param basic /usr/bin/squid-auth-engine /etc/squid-auth-engine.conf

And you're ready to use this module.

If you're a developer, you might be interested in reading through the source
code of this module, in order to learn about it's internals and how it works.
It may give you ideas about how to implement other plugin modules for
L<Net::Squid::Auth::Engine>. 

=head1 FUNCTIONS

=head2 new( $config_hash )

Constructor. Expects a hash reference with all the configuration under the
section I<< <SimpleLDAP> >> in the C<$Config{InstallScript}/squid-auth-engine> 
as parameter. Returns a plugin instance.

=cut

sub new {
    my ( $class, $config ) = @_;

    my $reftype = reftype($config) || '';
    croak 'Must pass a config hash' unless $reftype eq 'HASH';

    # some reasonable defaults
    $config->{userattr} = 'cn' unless $config->{userattr};
    $config->{passattr} = 'userPassword'
      unless $config->{passattr};
    $config->{objclass} = 'person' unless $config->{objclass};

    # required information
    foreach my $required qw(binddn bindpw basedn server) {
        croak qq{Missing config parameter '$required'}
          unless $config->{$required};
    }

    return bless { _cfg => $config }, $class;
}

=head2 initialize()

Initialization method called upon instantiation. This provides an opportunity 
for the plugin initialize itself, stablish database connections and ensure it 
have all the necessary resources to verify the credentials presented. It 
receives no parameters and expect no return values.

=cut

sub initialize {
    my $self = shift;

    # connect
    $self->{ldap} =
         Net::LDAP->new( $self->config('server'), $self->config('NetLDAP') )
      || croak "Cannot connect to LDAP server: " . $self->config()->{server};

    # bind
    my $mesg =
      $self->{ldap}
      ->bind( $self->config('binddn'), password => $self->config('bindpw') );
    $mesg->code && croak "Error binding to LDAP server: " . $mesg->error;

    return;
}

=head2 _search()

Searches the LDAP server. It expects one parameter with a search string for
the username. The search string must conform with the format used in LDAP
queries, as defined in section 3 of RFC 4515.

=cut

sub _search {
    my ( $self, $search ) = @_;

    # search
    my $mesg = $self->{ldap}->search(
        base   => $self->config('basedn'),
        scope  => 'sub',
        filter => '(&(objectClass='
          . $self->config('objclass') . ')('
          . $self->config('userattr') . '='
          . "$search" . '))',
        attrs => [ $self->config('userattr'), $self->config('passattr') ],
    );

    # if errors
    if ( $mesg->code ) {
        $mesg = $self->{ldap}->unbind;
        $mesg->code && croak "Error searching LDAP server: " . $mesg->error;
    }

    # get results
    my @entries = $mesg->entries();
    my $result  = {};

    my $entry = shift @entries;
    return $result unless $entry;

    my $user;
    if ( $self->config('userattr') =~ m/dn/i ) {
        $user = $entry->dn();
    }
    else {
        $user = $entry->get_value( $self->config('userattr') );
    }
    my $pw = $entry->get_value( $self->config('passattr') );

    $result->{$user} = $pw;

    carp "Found more than 1 entry for user ($user)" if shift @entries;

    return $result;
}

=head2 is_valid( $username, $password )

This is the credential validation interface. It expects a username and password
as parameters and returns a boolean indicating if the credentials are valid
(i.e., are listed in the configuration file) or not.

=cut

sub is_valid {
    my ( $self, $username, $password ) = @_;
    my $result = $self->_search("$username");
    return 0 unless exists $result->{$username};

    return $result->{$username} eq $password;
}

=head2 config( $key )

Accessor for a configuration setting given by key.

=cut

sub config {
    my ( $self, $key ) = @_;

    return $self->{_cfg}->{$key};
}

=head1 AUTHOR

Alexei Znamensky, C<< <russoz at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to 
C<bug-net-squid-auth-plugin-simpleldap at rt.cpan.org>, or through 
the web interface at 
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Net-Squid-Auth-Plugin-SimpleLDAP>.  
I will be notified, and then you'll automatically be notified of progress on 
your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Net::Squid::Auth::Plugin::SimpleLDAP

Or take a look at the github site to be up to date:

=over 4

L<http://github.com/russoz/Net-Squid-Auth-Plugin-SimpleLDAP>

=back

You can also look for information at:

=over 4

=item * RFC 4515 - Lightweight Directory Access Protocol (LDAP): String Representation of Search Filters

L<http://www.faqs.org/rfcs/rfc4515.html>

=item * RFC 4519 - Lightweight Directory Access Protocol (LDAP): Schema for User Applications

L<http://www.faqs.org/rfcs/rfc4519.html>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Net-Squid-Auth-Plugin-SimpleLDAP>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Net-Squid-Auth-Plugin-SimpleLDAP>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Net-Squid-Auth-Plugin-SimpleLDAP>

=item * Search CPAN

L<http://search.cpan.org/dist/Net-Squid-Auth-Plugin-SimpleLDAP>

=back

=head1 SEE ALSO

L<Net::Squid::Auth::Engine>, L<Net::LDAP>, L<Scalar::Util>

=head1 ACKNOWLEDGEMENTS

Luis "Fields" Motta Campos C<< <lmc at cpan.org> >>, who could now say:

"The circle is now complete. When I left you, I was but the learner; now *I* am the master."

To what I'd reply:

"Only a master of Perl, Fields"


=head1 COPYRIGHT & LICENSE

Copyright 2008,2010 Alexei Znamensky, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

42;    # End of Net::Squid::Auth::Plugin::SimpleLDAP

